conans/client/tools/env.py

```diff
@@ -116,31 +118,65 @@ def remove_from_path(command):
 
 
 def env_diff(cmd, only_diff):
-    known_path_lists = ("include", "lib", "libpath", "path")
-    if platform.system() == "Windows":
-        cmd += " && set"
-    else:
-        cmd += " && export"
+    # portable variant of shell/export command, as export may have different results, e.g.:
+    # USER=sse4 (zsh)
+    # declare -x USER="sse4" (bash)
+    # or may not be available at all! (csh)
+    # output as dictionary, as values may contain spaces, newlines, etc.
+    uuid_str = uuid.uuid4().hex  # needed to reliably find the start of the script output
+    get_env = 'from __future__ import print_function; import os; import json; ' \
+              'print(\'%s\'); print(json.dumps(dict(os.environ)))' % uuid_str
+    known_path_lists = (
+        "INCLUDE",
+        "LIB",
+        "LIBPATH",
+        "PATH",
+        "CPATH",
+        "C_INCLUDE_PATH",
+        "CPLUS_INCLUDE_PATH",
+        "OBJC_INCLUDE_PATH",
+        "FPATH",
+        "MANPATH",
+        "PYTHONPATH",
+        "CLASSPATH",
+        "LD_LIBRARY_PATH",
+        "DYLD_LIBRARY_PATH",
+        "DYLD_FALLBACK_LIBRARY_PATH",
+        "DYLD_VERSIONED_LIBRARY_PATH",
+        "DYLD_FRAMEWORK_PATH",
+        "DYLD_FALLBACK_FRAMEWORK_PATH",
+        "DYLD_VERSIONED_FRAMEWORK_PATH",
+        "DYLD_ROOT_PATH",
+        "PERL5LIB",
+        "GCONV_PATH",
+        "LOCPATH",
+        "NIS_PATH,"
+        "NLSPATH",
+        "GOPATH"
+    )
+    cmd += " && \"%s\" -c \"%s\"" % (sys.executable, get_env)
     ret = check_output_runner(cmd)
+    ret = ret[ret.find(uuid_str) + len(uuid_str):]
     new_env = {}
-    for line in ret.splitlines():
-        line = line.strip()
 
-        if line == "\n" or not line:
-            continue
+    # environment variables and paths are case-insensitive on Windows, case-sensitive otherwise
+    if platform.system() == "Windows":
+        def norm(x): return x.upper()
+    else:
+        def norm(x): return x
+
+    for name_var, value in json.loads(ret).items():
         try:
-            name_var, value = line.split("=", 1)
-            name_var = str(name_var)
-            value = str(value)
-            new_value = value.split(os.pathsep) if name_var.lower() in known_path_lists else value
+            new_value = value.split(
+                os.pathsep) if norm(name_var) in known_path_lists else value
             # Return only new vars & changed ones, but only with the changed elements if the var is
             # a list
             if only_diff:
                 old_value = os.environ.get(name_var)
-                if name_var.lower() == "path":
-                    old_values_lower = [v.lower() for v in old_value.split(os.pathsep)]
+                if norm(name_var) in known_path_lists:
```

Comment:
I am not sure why this ``env_diff`` needs to do UpperCase for that whole list of env-vars. I understand that the previous approach, coming from the ``vcvars`` implementation tries to workaround some PATH env-var issues, but I am a bit concerned about changing vars like GOPATH, PERL5LIB or GCONV_PATH. 

Also the vcvars was doing lowercase, while this one is upper-casing them, why the difference in behavior?

Reply:
My primary use of this is through compilervars, which can implicitly call vcvars, so the PATH env-var issues are still relevant here. Additionally, compilervars on linux will set things like LD_LIBRARY_PATH, MANPATH, NLSPATH and remove duplicate entries in PATH and LD_LIBRARY_PATH which can potentially cause issues if there were purposefully duplicate entries in there (for order considerations).

Upper vs lower doesn't matter as Windows is case sensitive, and the variables in known_path_lists are all uppercase, hence the use of uppercase.